### DISCUSSION TITLE
How to update the session when sending a request from a server component?

### DISCUSSION THREAD

**Discussion by mmoz** [vvo/iron-session#771] [Q&A] [answered]
In my app router, when I send a request from a server component and `console.log(session)`  in the route handler obtained from 
`const session = await getIronSession<SessionData>(cookies(), sessionOptions);`
 I get an empty object {} and I can't update the session. I would like to know if there is a way to update the session when sending a request from a server component?

**Comment by vvo** [comment]
@mmoz Can you create a very simple/small GitHub repository with your issue? So we can have a look, thanks!

**Comment by vvo** [comment, marked as answer]
@mmoz A server component cannot write cookies. That is a Next.js rule rather than an iron-session one: `cookies().set()` only works in a Server Action, a Route Handler, or `proxy.ts`. Reading a session in a server component is fine, but `session.save()` there has nowhere to go.

There are three places you can write one. In a Server Action or Route Handler, use `getIronSession(await cookies(), options)` directly. In Node or Express, use `getIronSession(req, res, options)`. And in `proxy.ts` (`middleware.ts` before Next 16), use the `nextProxyCookies` adapter that v9 adds, which is what you want when the session should be refreshed before the page renders. Saving in middleware did not work at all before v9.

On the empty `{}`: when a cookie exists but cannot be read, iron-session starts a fresh session rather than throwing. v9 lets you see why with `onUnsealError`, which would have told you whether the cookie was missing, expired, or sealed with a different password.

v9 notes: https://github.com/vvo/iron-session/releases/tag/v9.0.0
